From 004c1ec167b7a3061214a22ff29848530213f7c7 Mon Sep 17 00:00:00 2001
From: GH Cao <driver1998.ms@outlook.com>
Date: Fri, 18 Mar 2022 00:12:08 +0800
Subject: [PATCH 1/2] Introduce MSW ARM64 support

This is a preliminary ARM64 platform support for wxWidgets at "it
compiles" stage. This will allow building and testing wxWidgets based
apps for oncoming Windows 10 ARM64.

Backports from f69dbaa1aebd72652dfb6246bdeaffce4d83933d,
with MSVC-specific changes stripped.
---
 include/wx/msw/debughlp.h |  2 +-
 src/msw/stackwalk.cpp     | 10 ++++++++++
 2 files changed, 11 insertions(+), 1 deletion(-)

diff --git a/include/wx/msw/debughlp.h b/include/wx/msw/debughlp.h
index d0d289e..1ecfc92 100644
--- a/include/wx/msw/debughlp.h
+++ b/include/wx/msw/debughlp.h
@@ -185,7 +185,7 @@ public:
     // suffix in some cases. These 2 helper macros call the macro with the
     // correct arguments in both cases.
     #define wxSYM_CALL(what, name)  what(name, name)
-#if defined(_M_AMD64)
+#if defined(_M_AMD64) || defined(_M_ARM64)
     #define wxSYM_CALL_64(what, name)  what(name, name ## 64)
 
     // Also undo all the "helpful" definitions done by imagehlp.h that map 32
diff --git a/src/msw/stackwalk.cpp b/src/msw/stackwalk.cpp
index f9c8ca9..e510552 100644
--- a/src/msw/stackwalk.cpp
+++ b/src/msw/stackwalk.cpp
@@ -261,6 +261,16 @@ void wxStackWalker::WalkFrom(const CONTEXT *pCtx, size_t skip, size_t maxDepth)
     sf.AddrFrame.Mode      = AddrModeFlat;
 
     dwMachineType = IMAGE_FILE_MACHINE_AMD64;
+#elif defined(_M_ARM64)
+    // TODO: Verify this code once Windows 10 for ARM64 is commercially available
+    sf.AddrPC.Offset       = ctx.Pc;
+    sf.AddrPC.Mode         = AddrModeFlat;
+    sf.AddrStack.Offset    = ctx.Sp;
+    sf.AddrStack.Mode      = AddrModeFlat;
+    sf.AddrFrame.Offset    = ctx.Fp;
+    sf.AddrFrame.Mode      = AddrModeFlat;
+
+    dwMachineType = IMAGE_FILE_MACHINE_ARM64;
 #elif  defined(_M_IX86)
     sf.AddrPC.Offset       = ctx.Eip;
     sf.AddrPC.Mode         = AddrModeFlat;
-- 
2.35.1

